### PR TITLE
feat(auth): Add API to look up users by federated ID.

### PR DESCRIPTION
**Pull request by nrsim**

### CODE REVIEW

**Comment by hiranya911** [comment]
In general we don't add APIs to this SDK that are not already present in the Node.js SDK. 

@bojeil-google WDYT?

**Comment by bojeil-google** [comment]
Hey @hiranya911, I plan to review the Node.js implementation these 2 weeks. It's in my backlog. I think @rsgowman's plan was to launch multiple languages together (including the Java implementation).

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 1]
```diff
@@ -140,37 +141,98 @@ public void testDeleteNonExistingUser() throws Exception {
...
+    try {
+      UserRecord user1 = createUser(
+          randomUser1.uid, /* phoneNumber= */ null, randomUser1.email);
+      UserRecord user2 = createUser(
+          randomUser2.uid, randomPhoneNumber2, randomUser2.email);
+      UserImportResult user3 = importUser(
+          randomUser3.uid, randomPhoneNumber3, randomUser3.email, "google.com");
+
+      UserRecord lookedUpRecord = auth.getUserByPhoneNumberAsync(randomPhoneNumber2).get();
+      assertEquals(lookedUpRecord.getUid(), randomUser2.uid);
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: nit: junit uses `assertWhatever(expected, actual)`, so I think that `assertEquals(randomUser2.uid, lookedUpRecord.getUid())` would be preferred here. Throughout.

**Comment by hiranya911** [comment]
Sounds good. I'll assign to @rsgowman for now. Do we have an API proposal for this?

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 2]
```diff
@@ -140,37 +141,98 @@ public void testDeleteNonExistingUser() throws Exception {
...
     } finally {
-      auth.deleteUserAsync(userRecord.getUid()).get();
+      auth.deleteUserAsync(randomUser.uid).get();
+    }
+  }
+
+  @Test
+  public void testLookupUserByPhone() throws Exception {
+    RandomUser randomUser1 = RandomUser.create();
+    String phoneNumber1 = null;
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: nit: dead code. Either remove it or use it on line 179. (Personally, I'd lean toward removing it.)

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 3]
```diff
@@ -140,37 +141,98 @@ public void testDeleteNonExistingUser() throws Exception {
...
+      auth.deleteUserAsync(randomUser1.uid).get();
+      auth.deleteUserAsync(randomUser2.uid).get();
+      auth.deleteUserAsync(randomUser3.uid).get();
+    }
+  }
+
+  @Test
+  public void testLookupUserByFederatedId() throws Exception {
+    RandomUser randomUser1 = RandomUser.create();
+    String phoneNumber1 = null;
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: (again)

**Comment by bojeil-google** [comment]
Yeah, link is at: https://docs.google.com/document/d/1OfN5XH6cYJ7y3AnNsocftQSrbMiMaLlPlkWWWsOpDQg/

The proposal is approved.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 4]
```diff
@@ -140,37 +141,98 @@ public void testDeleteNonExistingUser() throws Exception {
...
+    try {
+      UserRecord user1 = createUser(
+          randomUser1.uid, /* phoneNumber= */ null, randomUser1.email);
+      UserRecord user2 = createUser(
+          randomUser2.uid, randomPhoneNumber2, randomUser2.email);
+      UserImportResult user3 = importUser(
+          randomUser3.uid, randomPhoneNumber3, randomUser3.email, "google.com");
+
+      UserRecord lookedUpRecord = auth.getUserByFederatedIdAsync(
+          randomUser3.uid + "_google.com", "google.com").get();
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: It's not clear why `randomUser3.uid + "_google.com"` is correct. Maybe add that as an explicit paramter to the importUser method? i.e. (on line 210-211):

```java
UserImportResult user3 = importUser(
    randomUser3.uid, randomPhoneNumber3, randomUser3.email,
    "google.com", randomUser3.uid + "_google.com");
```

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 5]
```diff
@@ -140,37 +141,98 @@ public void testDeleteNonExistingUser() throws Exception {
...
+      assertTrue(providers.contains("phone"));
+      assertTrue(providers.contains("google.com"));
+
+      try {
+        // Verify that lookup by federated identifier does not accept "phone".
+        lookedUpRecord = auth.getUserByFederatedIdAsync(
+            randomPhoneNumber3, "phone").get();
+        fail("No error thrown for non-federated provider");
+      } catch (IllegalArgumentException ignored) {
+        // expected
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: nit: You could rename the exception s/ignored/expected and then get rid of the comment.
http://go/java-style#s6.2-caught-exceptions

**Comment by hiranya911** [comment]
That only has the bulk get and delete API as far as I can tell. I don't see `getUserByFederatedId()` in there.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 6]
```diff
@@ -590,6 +652,52 @@ public void testGenerateSignInWithEmailLink() throws Exception {
     }
   }
 
+  private UserRecord createUser(
+      String uid,
+      @Nullable String phoneNumber,
+      @Nullable String email) throws Exception {
+    RandomUser randomUser = RandomUser.create();
```
Comment (rsgowman) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: Unused variable. rm?

(Rabbit-hole:  I would've figured that javac would warn about this; apparently not. But before I figured that out, I noticed that we don't have warnings enabled in this repo. Enabling them yields 116 warnings messages. Yikes. I haven't looked at them, so have no idea if any are actually interesting or not.)

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 7]
```diff
@@ -605,6 +605,56 @@ protected UserRecord execute() throws FirebaseAuthException {
     };
   }
 
+  /**
+   * Gets the user data corresponding to the specified user federated identifier.
```
Comment (rsgowman) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java]: > ...the specified user<b>'s</b> federated identifier.

(also below)

**Comment by rsgowman** [comment]
This is somewhat related to the bulk get, but is slightly different. The API doc for this feature is here: http://doc/1ikY2GoQZrwHNJ5mBa---t2o9R-MdjA4np4uW9SJ9dTg. It isn't approved yet (and thus we can't land this, but we don't want to delay starting on it while waiting for the API approval... especially this time of year.)

**Comment by bojeil-google** [comment]
Oh I didn't know that. I don't understand why this is being implemented before the other then. Definitely let's hold on this.

**Comment by nrsim** [comment]
We'll work on the Node implementation then. Sorry, I didn't know starting with the Node implementation was a must. I started with the Java implementation as the Node implementation was going to be tangled with bulk add/delete changes that are sent for review.

**Comment by bojeil-google** [comment]
Note I won't have time to review this PR. I will only have time to review the Node.js implementation. That should be the case going forward. @rsgowman and @hiranya911 reviews should be sufficient.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 8]
```diff
@@ -140,37 +141,92 @@ public void testDeleteNonExistingUser() throws Exception {
...
+      try {
+        // Verify that lookup by federated identifier does not accept "phone".
+        lookedUpRecord = auth.getUserByFederatedIdAsync(
+            randomUser3.phone, "phone").get();
+        fail("No error thrown for non-federated provider");
+      } catch (IllegalArgumentException expected) {
+      }
     } finally {
-      auth.deleteUserAsync(userRecord.getUid()).get();
+      auth.deleteUserAsync(randomUser1.uid).get();
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: We should use bulk delete for this when it becomes available. Add a TODO so we don't forget.

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 9]
```diff
@@ -605,6 +605,56 @@ protected UserRecord execute() throws FirebaseAuthException {
...
+   *
+   * @param providerUid The user identifier with the given provider.
+   * @param providerId Identifier for the given federated provider, for example,
+   *     "google.com" for the Google provider.
+   * @return A {@link UserRecord} instance.
+   * @throws IllegalArgumentException If the providerUid is null or empty, or if
+   *     the providerId is null, empty, or does not belong to a federated provider.
+   * @throws FirebaseAuthException If an error occurs while retrieving user data.
+   */
+  public UserRecord getUserByFederatedId(
```
Comment (rsgowman) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java:619]: Should be getUserByProviderUid(). (Though I think originally, this name was correct.) Also below (getUserByProviderUidAsync)

File: src/main/java/com/google/firebase/auth/AbstractFirebaseAuth.java [hunk 10]
```diff
@@ -553,6 +553,64 @@ protected UserRecord execute() throws FirebaseAuthException {
...
+   *     do not correspond to a user, the future throws a {@link FirebaseAuthException}.
+   * @throws IllegalArgumentException If the uid is null or empty, or if
+   *     the provider ID is null, empty, or does not belong to a federated provider.
+   */
+  public ApiFuture<UserRecord> getUserByProviderUidAsync(
+      @NonNull String providerId, @NonNull String uid) {
+    return getUserByFederatedIdOp(providerId, uid).callAsync(firebaseApp);
+  }
+
+  private CallableOperation<UserRecord, FirebaseAuthException> getUserByFederatedIdOp(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/AbstractFirebaseAuth.java:589]: Shouldn't this be called `getUserByProviderUidOp()` to be consistent with the style used in this class?

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 11]
```diff
@@ -145,6 +145,15 @@ UserRecord getUserByPhoneNumber(String phoneNumber) throws FirebaseAuthException
     return results;
   }
 
+  UserRecord getUserByFederatedId(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseUserManager.java:148]: getUserByProviderUid to be consistent with the terminology used in public API?

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 12]
```diff
@@ -265,6 +265,63 @@ public void testCreateUserWithParams() throws Exception {
     checkRecreateUser(randomUser.getUid());
   }
 
+  @Test
+  public void testLookupUserByPhone() throws Exception {
+    RandomUser randomUser1 = UserTestUtils.generateRandomUserInfo();
+    RandomUser randomUser2 = UserTestUtils.generateRandomUserInfo();
+    RandomUser randomUser3 = UserTestUtils.generateRandomUserInfo();
+    try {
+      createUser(randomUser1.getUid(), /* phoneNumber= */ null, randomUser1.getEmail());
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:274]: Don't seem to be used for anything.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 13]
```diff
@@ -265,6 +265,63 @@ public void testCreateUserWithParams() throws Exception {
...
+      assertEquals(randomUser3.getUid(), lookedUpRecord.getUid());
+    } finally {
+      auth.deleteUserAsync(randomUser1.getUid()).get();
+      auth.deleteUserAsync(randomUser2.getUid()).get();
+      auth.deleteUserAsync(randomUser3.getUid()).get();
+    }
+  }
+
+  @Test
+  public void testLookupUserByFederatedId() throws Exception {
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:293]: ProviderUid

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 14]
```diff
@@ -265,6 +265,63 @@ public void testCreateUserWithParams() throws Exception {
...
+    RandomUser randomUser1 = UserTestUtils.generateRandomUserInfo();
+    RandomUser randomUser2 = UserTestUtils.generateRandomUserInfo();
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:295]: user1 and user2 not used in the test.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 15]
```diff
@@ -929,6 +986,53 @@ public void onSuccess(ListProviderConfigsPage<SamlProviderConfig> result) {
     assertNull(error.get());
   }
 
+  private UserRecord createUser(
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:989]: You can possibly change this into `UserRecord createRandomUser()`. In the body you can use `RandomUser` to generate random data. That's a much simpler signature than passing 3 arguments.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 16]
```diff
@@ -929,6 +986,53 @@ public void onSuccess(ListProviderConfigsPage<SamlProviderConfig> result) {
...
+      user.setPhoneNumber(phoneNumber);
+    }
+    if (email != null) {
+      user.setEmail(email);
+      user.setEmailVerified(true);
+    }
+    return auth.createUserAsync(user).get();
+  }
+
+  private UserImportResult importUser(
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:1008]: Same here.

```
UserRecord importRandomUser(String providerId, String providerUid)
```

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Implementation looks pretty solid. Just a few thoughts on cleaning up the tests a bit.

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java [hunk 17]
```diff
@@ -267,45 +267,37 @@ public void testCreateUserWithParams() throws Exception {
...
 
-      lookedUpRecord = auth.getUserByPhoneNumberAsync(randomUser3.getPhoneNumber()).get();
-      assertEquals(randomUser3.getUid(), lookedUpRecord.getUid());
+      lookedUpRecord = auth.getUserByPhoneNumberAsync(user2.getPhoneNumber()).get();
+      assertEquals(user2.getUid(), lookedUpRecord.getUid());
     } finally {
-      auth.deleteUserAsync(randomUser1.getUid()).get();
-      auth.deleteUserAsync(randomUser2.getUid()).get();
-      auth.deleteUserAsync(randomUser3.getUid()).get();
+      if (user1 != null) {
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseAuthIT.java:283]: I just remembered that we now have a `TemporaryUser` abstraction that handles automatic clean up of user accounts created during testing. Can we tie `createRandomUser()` and `importRandomUser()` methods to that abstraction, and clean up this logic?


Here's an example of how it's being used:
https://github.com/firebase/firebase-admin-java/blob/d8b1583002d60568106bf4a7ba2d5bcbbb6c0463/src/test/java/com/google/firebase/auth/FirebaseAuthIT.java#L446-L449

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Looks pretty solid @rsgowman. Just one more comment concerning the integration tests (Sorry for not bringing that up earlier, but I completely forgot about the `TemporaryUser` abstraction that we now use in our tests).

File: src/test/java/com/google/firebase/auth/FirebaseAuthIT.java (hunk 17 above)
Comment (rsgowman) [reply src/test/java/com/google/firebase/auth/FirebaseAuthIT.java]: Done.

**Review by hiranya911 — APPROVED** [review]
LGTM 👍 

Let me know when you want this merged, and I can perform a force merge to override the CLA bot.

**Pull request merged**